File: src/brs/Blockchain.java
package brs;

import brs.util.CollectionWithIndex;
import java.util.Collection;

//TODO: Create JavaDocs and remove this
@SuppressWarnings({ "checkstyle:MissingJavadocTypeCheck", "checkstyle:MissingJavadocMethodCheck" })

public interface Blockchain {

    Block getLastBlock();

    Block getLastBlock(int timestamp);

    void setLastBlock(Block blockImpl);

    int getHeight();

    Block getBlock(long blockImplId);

    Block getBlockAtHeight(int height);

    boolean hasBlock(long blockImplId);

    Collection<Block> getBlocks(int from, int to);

    Collection<Block> getBlocks(Account account, int timestamp);

    CollectionWithIndex<Block> getBlocks(Account account, int timestamp, int from, int to);

    int getBlocksCount(long accountId, int from, int to);

    Collection<Long> getBlockIdsAfter(long blockImplId, int limit);

    Collection<Block> getBlocksAfter(long blockId, int limit);

    long getBlockIdAtHeight(int height);

    Transaction getTransaction(long transactionId);

    Transaction getTransactionByFullHash(String fullHash); // TODO add byte[] method

    boolean hasTransaction(long transactionId);

    boolean hasTransactionByFullHash(String fullHash); // TODO add byte[] method

    int getTransactionCount();

    Collection<Transaction> getAllTransactions();

    public long getAtBurnTotal();

    public long getTotalMined();

    long getBlockReward(int height);

    CollectionWithIndex<Transaction> getTransactions(
            Long senderId,
            Long recipientId,
            int numberOfConfirmations,
            byte type,
            byte subtype,
            int blockTimestamp,
            int from,
            int to,
            boolean includeIndirectIncoming,
            boolean bidirectional);

    CollectionWithIndex<Transaction> getTransactions(
            Account account,
            int numberOfConfirmations,
            byte type,
            byte subtype,
            int blockTimestamp,
            int from,
            int to,
            boolean includeIndirectIncoming);

    Collection<Transaction> getTransactions(
            long senderId,
            byte type,
            byte subtypeStart,
            byte subtypeEnd,
            int from,
            int to);

    int countTransactions(byte type, byte subtypeStart, byte subtypeEnd);

    Collection<Transaction> getTransactionsWithFullHashReference(
            String fullHash,
            int numberOfConfirmations,
            byte type,
            byte subtypeStart,
            byte subtypeEnd,
            int from,
            int to);

    Collection<Long> getTransactionIds(
            Long sender,
            Long recipient,
            int numberOfConfirmations,
            byte type,
            byte subtype,
            int blockTimestamp,
            int from,
            int to,
            boolean includeIndirectIncoming);

    long getCommittedAmount(long accountId, int height, int endHeight, Transaction skipTransaction);
}

File: src/main/java/brs/Blockchain.kt
package brs

interface Blockchain {

    var lastBlock: Block

    val height: Int

    fun getTransactionCount(): Int

    fun getAllTransactions(): Collection<Transaction>

    fun getLastBlock(timestamp: Int): Block?

    fun getBlock(blockId: Long): Block?

    fun getBlockAtHeight(height: Int): Block?

    fun hasBlock(blockId: Long): Boolean

    fun getBlocks(from: Int, to: Int): Collection<Block>

    fun getBlocks(account: Account, timestamp: Int): Collection<Block>

    fun getBlocks(account: Account, timestamp: Int, from: Int, to: Int): Collection<Block>

    fun getBlockIdsAfter(blockId: Long, limit: Int): Collection<Long>

    fun getBlocksAfter(blockId: Long, limit: Int): Collection<Block>

    fun getBlockIdAtHeight(height: Int): Long

    fun getTransaction(transactionId: Long): Transaction?

    fun getTransactionByFullHash(fullHash: ByteArray): Transaction?

    fun hasTransaction(transactionId: Long): Boolean

    fun hasTransactionByFullHash(fullHash: ByteArray): Boolean

    fun getTransactions(account: Account, type: Byte, subtype: Byte, blockTimestamp: Int, includeIndirectIncoming: Boolean): Collection<Transaction>

    fun getTransactions(account: Account, numberOfConfirmations: Int, type: Byte, subtype: Byte, blockTimestamp: Int, from: Int, to: Int, includeIndirectIncoming: Boolean): Collection<Transaction>

    fun setLastBlock(previousBlock: Block, block: Block)
}
